actioncable/app/javascript/action_cable/logger.js

```diff
@@ -0,0 +1,18 @@
+import adapters from "./adapters"
+
+let enabled
+
+export function log(...messages) {
+  if (enabled) {
+    messages.push(Date.now())
+    adapters.logger.log("[ActionCable]", ...messages)
+  }
+}
+
+export function startDebugging() {
+  enabled = true
+}
+
+export function stopDebugging() {
+  enabled = false
+}
```

Comment:
I don't think `{start,stop}Debugging()` make sense in the context of a logger. What do you think about exporting an object with a `log()` method and `enabled` property? That way you can toggle logging with `logger.enabled = true|false`, and log with `logger.log(…)`.

Reply:
I think we could do that but it wouldn't minify as well as this currently does, because the minifier can't freely mangle the exposed property name `enabled` to `e` like it can with a variable name since it knows that variable is only used internally. And when you import from the source code, `startDebugging` and `stopDebugging` are also just variable names rather than properties, so they can also be mangled to single letters, or removed entirely when the application doesn't call those functions.

With that said, `startDebugging` and `stopDebugging` are effectively `enableLogging` and `disableLogging`, which is why I grouped them in the logger module. What if we renamed them to `enableLogging` and `disableLogging`, instead? That's more clear and wouldn't have to sacrifice the mangling benefits.

Reply:
It's not worth sacrificing the API over literally a few bites. I'd like a logger object to be exported from `index.js`, not each of its methods. The API should either be `logger.enabled = true|false` or `logger.enable()`/`logger.disable()`.

Reply:
I'm cool with `logger.enable()`/`logger.disable()` 👍

Reply:
Ok, the API is now `ActionCable.logger.enable()/disable()`

A separate thing is that it's not ideal that we're currently _always_ shipping call's to `ActionCable.log` inside ActionCable's methods, for example:
https://github.com/rails/rails/blob/604fac6d7191fca102380b1a5f5eb9c619fb407f/actioncable/app/javascript/action_cable/connection.js#L29-L34

Those `log(...)` calls are always there. So even if you never call `ActionCable.logger.enable()`, your production build contains all those `log` calls. But not only does that mean your bundle ends up bigger than it needs to (slowing the download time), but it also slows the runtime execution - each `log` call will end up as a no-op, but it still has to evaluate the arguments, which as seen in the example above can call other methods like `getState()`, etc., despite never using the result).

Since `log` is mainly useful for development, it would be great if there was a way to have these calls stripped from production builds to avoid both download time and runtime bloat.

Reply:
> Since `log` is mainly useful for development

FWIW, it's only been useful for us in production. Flipping on the logger when something starts acting up helped us diagnose a number of issues. That said, we haven't used it in a while now that things have stabilized.

> it would be great if there was a way to have these calls stripped from production builds to avoid both download time and runtime bloat

I'm open to that. Let's explore it separately.